Src/PCompiler/CompilerCore/TypeChecker/ModuleExprVisitor.cs

```diff
@@ -30,34 +56,37 @@ public static void PopulateAllModuleExprs(
             // all the test declarations
             foreach (var test in globalScope.SafetyTests)
             {
-                if (test.ParamExpr == null)
+                switch (test.TestKind)
                 {
-                    var context = (PParser.ParametricSafetyTestDeclContext)test.SourceLocation;
-                    test.ModExpr = modExprVisitor.Visit(context.modExpr());
-                    var expr = (NamedTupleExpr)paramExprVisitor.Visit(context.globalParam);
-                    var names = ((NamedTupleType)expr.Type).Names.ToList();
-                    var values = expr.TupleFields.ToList();
-                    IDictionary<string, List<IPExpr>> dic = new Dictionary<string, List<IPExpr>>();
-                    foreach (var (name, value) in names.Zip(values))
+                    case TestKind.NormalTest:
                     {
-                        var v = globalScope.Variables.FirstOrDefault(x => x.Name == name);
-                        if (v == null)
-                        {
-                            throw handler.UndeclaredGlobalConstantVariable(context.globalParam, name);
-                        }
-                        var expectedType = new SequenceType(v.Type);
-                        if (!value.Type.Equals(expectedType))
-                        {
-                            throw handler.TypeMismatch(value.SourceLocation, value.Type, expectedType);
-                        }
-                        dic[name] = ((SeqLiteralExpr)value).Value;
+                        var context = (PParser.SafetyTestDeclContext)test.SourceLocation;
+                        test.ModExpr = modExprVisitor.Visit(context.modExpr());
+                        break;
+                    }
+                    case TestKind.AssumeParametricTest:
+                    {
+                        var context = (PParser.ParametricAssumeSafetyTestDeclContext)test.SourceLocation;
+                        test.ModExpr = modExprVisitor.Visit(context.modExpr());
+                        var expr = (NamedTupleExpr)paramExprVisitor.Visit(context.globalParam);
+                        var dic = make_assignments (handler, globalScope, context.globalParam, expr);
+                        test.ParamExpr = dic;
+                        test.AssumeExpr = exprVisitor.Visit(context.assumeExpr);
+                        break;
+                        
+                    }
+                    case TestKind.ParametricTest:
+                    {
+                        var context = (PParser.ParametricSafetyTestDeclContext)test.SourceLocation;
+                        test.ModExpr = modExprVisitor.Visit(context.modExpr());
+                        var expr = (NamedTupleExpr)paramExprVisitor.Visit(context.globalParam);
+                        var dic = make_assignments (handler, globalScope, context.globalParam, expr);
+                        test.ParamExpr = dic;
+                        // the default assumption is true 
```

Comment:
I have a strong feeling we can merge the AssumeParam and NormalParam together; making NormalParam a special case of AssumeParam where the assumes are all true?